### PR TITLE
New tutorial: Debugging Galaxy

### PR DESCRIPTION
**Pull request by ic4f**
Apologies for the numerous typos - will fix.
Any feedback very much appreciated!

TODO:
- [x] Address https://github.com/galaxyproject/training-material/pull/2614#discussion_r648442904 (point to specific link, see discusssion)
- [x] Address https://github.com/galaxyproject/training-material/pull/2614#discussion_r648445717 (yarn)
- [x] Address https://github.com/galaxyproject/training-material/pull/2614#discussion_r648462320 (explicit paths)
- [x] Address https://github.com/galaxyproject/training-material/pull/2614#discussion_r648458532 (selenium test issue)
- [x] replace todos in key points with key points

### CODE REVIEW

**Comment by ic4f** [comment]
OK, so now I remember why we were holding off with this PR! The key points are not done. dammit... We'll submit then in a separate commit.

**Comment by hexylena** [comment]
Not a problem :) this at least lets everyone else start looking and making suggestions but the formatting already looks quite good :)

File: topics/dev/tutorials/debugging/tutorial.md
Comment (davelopez) [inline topics/dev/tutorials/debugging/tutorial.md:422]: [The Selenium test that is failing](https://github.com/galaxyproject/galaxy/runs/2784473904) in the `training` branch is:
```
lib/galaxy_test/selenium/test_history_dataset_state.py::HistoryDatasetStateTestCase::test_dataset_change_dbkey
```
According to https://github.com/galaxyproject/galaxy/commit/8f1e14bed8a24da6117051f13f30cd96ff62318f it is probably correct, but then this tutorial block probably needs to be updated to address it instead of the proposed ones or vice-versa.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (davelopez) [inline topics/dev/tutorials/debugging/tutorial.md:442]: Maybe it's too obvious, but since in the last tutorial block we were in the `GALAXY_ROOT/client` directory, it may be worth remember that we should be now again in the `GALAXY_ROOT` dir to run the commands.

**Review by davelopez — COMMENTED** [review]
Excellent tutorial! Really easy to follow and enjoyable.

Only the Selenium block needs to be checked a bit in case there was some confusion with the tests that should fail in the `training` branch.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (davelopez) [reply topics/dev/tutorials/debugging/tutorial.md:312]: Oh! I just saw that you already tried that workaround in https://github.com/galaxyproject/galaxy/pull/12111

Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:312]: Yeah, it did it again. I think we need to point to the specific commit in the tutorial where that workflow (or build) did in fact run: https://github.com/galaxyproject/galaxy/runs/2780158706 .

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:324]: Hmm, I always took it for granted: I thought it was installed automagically! If not then yes, sure, we should include that!

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:442]: Good point. Maybe each time we refer to paths we should prepend GALAXY_ROOT - then the difference would be obvious.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:365]: It runs on my setup. Misconfig on your end?

Comment (davelopez) [reply topics/dev/tutorials/debugging/tutorial.md:365]: Most probably, but I didn't really set up anything apart from the steps of the tutorial. I didn't even have `yarn` installed in this machine as you can see :laughing:

Feel free to ignore this, if someone has a similar problem they have `yarn jest` as an alternative so, no problem.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (assuntad23) [inline topics/dev/tutorials/debugging/tutorial.md:421]: @ic4f This test is not failing... whoops? We should remove this line

File: topics/dev/tutorials/debugging/tutorial.md
Comment (assuntad23) [inline topics/dev/tutorials/debugging/tutorial.md:438]: And the corresponding file here...

File: topics/dev/tutorials/debugging/tutorial.md
Comment (assuntad23) [inline topics/dev/tutorials/debugging/tutorial.md:443]: @ic4f, I never give them a sample command to run in a Solution block... I'm certain that quick readers will simply think test_sign_out.py is what they're supposed to run and be confused when it all passes. Maybe we should explicitly state the command, as we do everywhere else.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (assuntad23) [reply topics/dev/tutorials/debugging/tutorial.md:312]: This is unfortunate...

File: topics/dev/tutorials/debugging/tutorial.md
Comment (assuntad23) [reply topics/dev/tutorials/debugging/tutorial.md:324]: I'm pretty sure (though, by no means an expert & have great potential to be wrong) that yarn is installed as part of the virtual environment, which would make it unnecessary to install it first.

Someone -- fact check me.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (assuntad23) [reply topics/dev/tutorials/debugging/tutorial.md:422]: Another solution, which is potentially not the best idea, is to just delete the selenium test that is failing here? 

Make it seem more like a runtime error that doesn't have any existing code coverage. ¯\_(ツ)_/¯

File: topics/dev/tutorials/debugging/tutorial.md
Comment (davelopez) [reply topics/dev/tutorials/debugging/tutorial.md:324]: You are right, the documentation says:

> The base dependencies used are Node.js and Yarn. Galaxy now includes these in
the virtual environment, and they can be accessed by activating that with `.
.venv/bin/activate` from the Galaxy root directory.

I don't know what went wrong then, I checked the tutorial in a freshly installed Ubuntu and I would have sworn that the virtual environment was activated (as I carefully followed every step) but, in my case, I have **huge** potential to be wrong :laughing:  so this can be safely ignored, as I said, is not a big deal and `make client-lint` points you to the installation instructions in case someone steps into it :)

File: topics/dev/tutorials/debugging/tutorial.md
Comment (davelopez) [reply topics/dev/tutorials/debugging/tutorial.md:443]: I think that's a great idea and makes the tutorial more reproducible and easy to follow :)

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:59]: maybe github actions?

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:324]: Thank you for checking and verifying!

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:443]: I agree, let's have explicit commands in solutions (otherwise users won't be able to verify, beyond any doubt, that what they did was right and not just coincidentally worked)

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:85]: to be honest, I am not sure if we need this section. all this can be created with `sh run.sh`
Do we really want them to configure virtualenv themselves?

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:59]: If this is more appropriate/easy to understand, yes, sure! (was my idea to call them workflows - and I was not at all sure that was a good term in this context). Feel free to suggest the necessary edits - there may be more than one instance.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:85]: It's much faster than `run.sh`, and we don't need `run.sh` to run tests in the first part of the tutorial. I think that'll get them going faster; besides, that may be yet another useful thing to learn - it might take the mystery out of .venv for some people, I think.

Comment (OlegZharkov) [reply topics/dev/tutorials/debugging/tutorial.md:85]: maybe (just to skip the installation part), we can show them how to run 1 random test with `run_tests.sh`? It should create .venv as well

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:331]:
```suggestion
>    > > In the file `client/src/components/RuleBuilder/RuleComponent`, there is an unused import. Simply delete the line that says `import Vue from "vue";` 
Alternatively, you can run `yarn eslint --fix`
```

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:343]:
```suggestion
5. __Little Hint__
 You also can auto-fix your linting problems with `yarn run prettier` and `yarn eslint --fix`
```

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:412]:
```suggestion
>    >  There may be actually a few failing Selenium tests, but the good news are, they are the result of only one little error. The failing tests on GitHub says "Selenium Test 3.7, 2". When you click on the details of this one, it's not always very clear which test is causing the failure, but if you download the artifacts, it's very clear which tests have failed.
```

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:452]:
```suggestion 
The problem is that the tests are expecting a different HID (history identifier) than what is being made available to them.  Since all the failing tests have to do with Lists, you might want to check out the ListCollectionCreator.
```

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [reply topics/dev/tutorials/debugging/tutorial.md:422]: > Another solution, which is potentially not the best idea, is to just delete the selenium test that is failing here?
> 
> Make it seem more like a runtime error that doesn't have any existing code coverage. ¯_(ツ)_/¯

I support this idea. It's not like we're developing something for production. Our goal here is to create a small usecase, deleting them will reduce confusion and that will contribute the teaching process.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:467]:
```suggestion
>    >  Excellent! If you're here, you've found the problem, ran it locally, fixed the problem, ran it locally again without errors, and are ready to commit your change. Write a useful commit message like `Fix logic error in ListCollectionCreator` or something similar, making sure it is clear what you've done.
```

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:44]:
```suggestion
In this tutorial we will demonstrate how to find and fix common types of bugs you may encounter as a contributor to Galaxy. We will step you through the process of finding and fixing a bug - from locating specific errors in the logs of Galaxy's GitHub Actions, to identifying their cause, developing a solution and committing your edits
```

File: topics/dev/tutorials/debugging/tutorial.md
Comment (OlegZharkov) [inline topics/dev/tutorials/debugging/tutorial.md:59]:
```suggestion
1. Locate failing test or error on GitHub Actions
```

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: Applied in separate commit.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: Applied, only "new is", not "are". (I know it feels like it should be plural, but the noun is uncountable, so "is")

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: Applied. Good point!

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: Done.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: I didn't prepend it to all paths - I thought it looked more confusing that way; but I did make it explicit throughout the tutorial: 6966e47593be42e208d4c27edccdcbfb4c45db36

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: Deleting a test is an option.. but that opens precendent: on a future update some other tests will fail due to our "installed" bugs - we would have to delete those too, etc.
If there are multiple failing tests and only some of them are relevant to a given bug, I think we can explain why there are to be expected multiple test failures - somewhere at the beginning of the tutorial (or at the beginning of that section). If the test is not available (i.e., it did not run due to some other bug we introduced), then we can just name that test.

I'm not opposed to deleting tests to accommodate the tutorial, but I think we should try to avoid this if we can.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:85]: We do ask them to use `run _tests.sh` for the api and selenium sections. Also, I've added this line that says that run_tests.sh will activate the venv (not create, since we ask them to create it at the beginning of the tutorial, so they already have it):
https://github.com/galaxyproject/training-material/pull/2614/files#diff-03c321ac89fe848061d63bb53e4d75eec9a2f914b4b13efa2f86afae057572a4R238

**Comment by hexylena** [comment]
I pushed a couple of formatting fixes just fyi. The main thing was a whitespace issue causing your numbering to reset in the setup instructions.

We also have the code-in / code-out classes for showing commands with their output very clearly. Since there are a lot of commands run at the start maybe it's more annoying than it's worth, but I thought I'd show it as an example, feel free to revert.

image

additionally I removed some of the prefixed `$` from commands as it makes it easier for people to copy and paste without it erroring.

**Comment by ic4f** [comment]
@hexylena thank you for the edits!
 
The solution wrapped in a Question box looks very nice (and logical).

I considered code-in/code-out, but, I think, that's the only place where we need to display output that is a result of code we've just typed in. But for this example it looks good and makes it clear, thanks!

I was torn about the prefix - other tutorials use it inconsistently. I decided to go with the prefix to distinguish between (a) code that a user is expected to type in at the command prompt versus (b) output from tests or Python code, etc. So it's a trade-off between simplifying copying and pasting -- and distinguishing between a command to be run and code to be read.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md:381]: Added as a tip to the linting section. Thanks!

**Comment by hexylena** [comment]
Ahh interesting to hear your rationale. That's a bit why we added the in/out boxes as well, to help make it clear what is something you type vs something that's an output.

I'd love to get rid of the `$` completely from every other tutorial because people really like to copy and paste based on feedback. That said, it's a more technical audience here, maybe they're used to that. And additionally there is the option to prefix with `$` in css which would prevent it being copied, but we'd probably need to annotate the code blocks where there's a command.

**Comment by ic4f** [comment]
@hexylena Thanks for the suggestion. At first I thought of in/out boxes only to be applicable to exampes of direct cause and effect - which is not often the case; but I never considered input only - and that seems fine. I've added more boxes for consistency.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: Handled in 9d055bcb6570fbf87cf3819bdf3684f618cb7709

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: This has been updated. I'll edit the wording too to make it clear that more than one workflow needs to be checked.

File: topics/dev/tutorials/debugging/tutorial.md
Comment (ic4f) [reply topics/dev/tutorials/debugging/tutorial.md]: already fixed - tests fail, so this is fine.

**Comment by ic4f** [comment]
OK, this should be done.

**Comment by hexylena** [comment]
I'll get it reviewed today.

**Review by hexylena — APPROVED** [review]
This looks fantastic @assuntad23 @ic4f @jmchilton. I really wish I'd had this tutorial when I was starting out as a galaxy contributor, it would've made my life so much better. I like the variety of error cases presented, and the improvements to the linting such that a number of steps are "read the error message, you should know how to fix it". Fantastic

I'm happy with the formatting, everything looks good to me. I'll merge when tests pass.

**Comment by hexylena** [comment]
Special thanks to @davelopez @OlegZharkov for the reviewing!

**Pull request merged**
